src/java/org/apache/cassandra/db/memtable/Memtable.java

```diff
@@ -0,0 +1,389 @@
+         * want to flush, it should return true here and false on shouldSwitch(STREAMING/REPAIR).
+         * In that case, temporary sstables will be created, streamed and then deleted.
+         */
+        default boolean streamFromMemtable()
+        {
+            return false;
+        }
+    }
+
+    /**
+     * Interface for providing signals back to the owner.
+     */
+    interface Owner
+    {
+        void signalFlushRequired(Memtable memtable, ColumnFamilyStore.FlushReason reason);
+    }
+
+
+    // Main write and read operations
+
+    /**
+     * Put new data in the memtable. This operation may block until enough memory is available in the memory pool.
+     *
+     * @param update the partition update, may be a new partition or an update to an existing one
+     * @param indexer receives information about the update's effect
+     * @param opGroup write operation group, used to permit the operation to complete if it is needed to complete a
+     *                flush to free space.
+     *
+     * @return the smallest timestamp delta between corresponding rows from existing and update. A
+     * timestamp delta being computed as the difference between the cells and DeletionTimes from any existing partition
+     * and those in {@code update}. See CASSANDRA-7979.
+     */
+    long put(PartitionUpdate update, UpdateTransaction indexer, OpOrder.Group opGroup);
+
+    /**
+     * Get the partition for the specified key. Returns null if no such partition is present.
+     */
+    Partition getPartition(DecoratedKey key);
+
+    /**
+     * Returns a partition iterator for the given data range.
+     *
+     * @param columnFilter filter to apply to all returned partitions
+     * @param dataRange the partition and clustering range queried
+     */
+    MemtableUnfilteredPartitionIterator makePartitionIterator(ColumnFilter columnFilter,
+                                                              DataRange dataRange);
+
+    interface MemtableUnfilteredPartitionIterator extends UnfilteredPartitionIterator
+    {
+        /**
+         * Returns the minimum local deletion time for all partitions in the range.
+         * Required for the efficiency of partition range read commands.
+         */
+        int getMinLocalDeletionTime();
+    }
+
+
+    // Statistics
+
+    /** Number of partitions stored in the memtable */
+    int partitionCount();
+
+    /** Size of the data not accounting for any metadata / mapping overheads */
+    long getLiveDataSize();
+
+    /** Number of operations (i.e. {@link #put} calls) the memtable has executed */
+    long getOperations();
+
+    /** Minimum timestamp of all stored data */
+    long getMinTimestamp();
+
+    /**
+     * The table's definition metadata.
+     *
+     * Note that this tracks the current state of the table and is not necessarily the same as what was used to create
+     * the memtable.
+     */
+    TableMetadata metadata();
+
+
+    // Memory usage tracking
+
+    /**
```

Comment:
I think that this API could be improved:

1. MemoryUsage is non-static, i.e. it's tied to some Memtable; yet, there is `addMemoryUsageTo`, which means that one memtable can add memory usage to another memtable; at the same time there is `getMemoryUsage` which is supposed to return the memory used by the memtable. But how is the memtable supposed to know it's memory usage if it must be partially injected with `addMemoryUsageTo`? It's not clear from the API what is the intended purpose.

This is very error prone. E.g. `getMemoryUsage` is really `createMemoryUsage`, but it's called `get`. How do we ensure that the implementation doesn't e.g. store one `MemoryUsage` object and just returns it on demand? In that case, how do we ensure that the same memory is not added twice? 
I read the implementation of the methods and I _think_ that the purpose can be expressed without being error-prone as:

- `MemoryUsage` is static (can be related to a memtable, or can be a free aggregate)
- `MemoryUsage` has a method for combining two `MemoryUsage` (`add` method?); 
- there should be an assertion that memory pools are the same in both `added` objects; if that assertion does not hold then it means that the MemoryUsage class layout is invalid, and the pool should be someplace else, perhaps in the Memtable itself.
- memtable has a `createMemoryUsage` method (or `newMemoryUsage` or whatever is the convention) which implementations are supposed to implement with "create" behavior
- `addMemoryUsageTo` is removed (it's not memtable responsibility to know how to combine `MemoryUsage` of two memtables); 

alternatively, a memtable could know "related" memtables and it would be a `getMemoryUsage` function that would combine the memtable's usage with the usages of all related memtables.  that's probably not feasible as we do not have this relationship on the level of memtable (and that's fine)

alternatively, MemoryUsage could be left as non-static, but the adding mechanism would have to be moved away from the `MemoryUsage` class; that's probably not feasible; the question is why the memory pool is in `MemoryUsage`

Reply:
kind of nit; consider, but not necessarily apply:

2. As a rule of the thumb I'd rather we separated interfaces used by different clients (interface segregation principle)
E.g. I don't see a point in having the same Memtable interface for writers (e.g. `put`) and memory trackers (e.g. `getMemoryUsage`). 
For one thing - it makes testing easier, as to test a specific client we do not need to be concerned with unrelated parts; 
For another - it makes any further extensions, refactorings and reuse easier.

That said, I admit I do not know how to segregate the interfaces correctly. However, some distinct interfaces seem to be already segretated with comments: "memory usage tracking", "main write operations", "main read operations", "lifecycle management", "flushing"
I also admit that the point about testing is kind of moot when we're dealing with lot's of legacy code, which sometimes won't allow us to unittest.
I realize that further extensions, refactorings and reuse is rather unlikely in this particular case.

However, in my experience establishing a practice of adhering to SOLID principles proved to be a very-low cost of preventing tech debt buildup while allowing rapid reconnecting of existing pieces to build new functionalities.

Reply:
I agree, this was hacky, and the main reason was the flush-triggering logic being in the wrong place. Moved it to `AbstractAllocatorMemtable` which makes it possible to make the usage object static.